lib/bolt/catalog.rb

```diff
@@ -65,23 +68,45 @@ def compile_catalog(request)
       target = request['target']
       pdb_client = Bolt::PuppetDB::Client.new(Bolt::PuppetDB::Config.new(request['pdb_config']))
       options = request['puppet_config'] || {}
+
       with_puppet_settings(request['hiera_config']) do
         Puppet[:rich_data] = true
         Puppet[:node_name_value] = target['name']
         Puppet::Pal.in_tmp_environment('bolt_catalog',
-                                       modulepath: request["modulepath"] || [],
-                                       facts: target["facts"] || {},
-                                       variables: target["variables"] || {}) do |pal|
+                                       modulepath: request['modulepath'] || [],
+                                       facts: target['facts'] || {},
+                                       variables: target['variables'] || {}) do |pal|
+          inv = if request['future']
+                  Bolt::ApplyInventory.new(request['config'])
+                else
+                  setup_inventory(request['inventory'])
+                end
           Puppet.override(bolt_pdb_client: pdb_client,
-                          bolt_inventory: setup_inventory(request['inventory'])) do
+                          bolt_inventory: inv) do
             Puppet.lookup(:pal_current_node).trusted_data = target['trusted']
             pal.with_catalog_compiler do |compiler|
+              if request['future']
+                # This needs to happen inside the catalog compiler so loaders are initialized for loading
+                vars = Puppet::Pops::Serialization::FromDataConverter.convert(request['plan_vars'])
+                pal.send(:add_variables, compiler.send(:topscope), vars.merge(target['vars']))
```

Comment:
We call it `target['vars']` in future but `target['variables']` in the old version. I see that that works around the fact that we would otherwise be setting the target variables earlier (`Puppet::Pal.in_tmp_environment`) and this avoids a conflict. There must be a clearer way to handle that though.